flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/RestartStreamingJobBenchmark.java

```diff
@@ -0,0 +1,96 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+package org.apache.flink.runtime.scheduler.benchmark.e2e;
+
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.concurrent.ComponentMainThreadExecutorServiceAdapter;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobmaster.slotpool.SlotPoolUtils;
+import org.apache.flink.runtime.scheduler.DefaultScheduler;
+
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.StreamSupport;
+
+/**
+ * The benchmark of restarting tasks in a STREAMING/BATCH job. The related method is {@link
+ * DefaultScheduler#handleGlobalFailure}.
+ */
+public class RestartStreamingJobBenchmark extends SchedulingAndDeployingBenchmark {
+    private static final int SLOTS_PER_TASK_EXECUTOR = 4;
+
+    @Override
+    protected void offerSlots() {
+        final int numberSlots =
+                StreamSupport.stream(jobGraph.getVertices().spliterator(), false)
+                        .mapToInt(JobVertex::getParallelism)
+                        .sum();
+
+        for (int i = 0; i < Math.ceil((double) numberSlots / SLOTS_PER_TASK_EXECUTOR); i++) {
+            SlotPoolUtils.tryOfferSlots(
+                    slotPool,
+                    ComponentMainThreadExecutorServiceAdapter.forMainThread(),
+                    Collections.nCopies(SLOTS_PER_TASK_EXECUTOR, ResourceProfile.ANY));
+        }
+    }
+
+    public void startSchedulingAndFailover() throws Exception {
+        startScheduling();
+
+        CompletableFuture.runAsync(
+                        () -> {
+                            // trigger failover and force reset state to cancelled.
+                            scheduler.handleGlobalFailure(new RuntimeException());
+                            scheduler
+                                    .getExecutionGraph()
+                                    .getAllExecutionVertices()
+                                    .forEach(
+                                            executionVertex ->
+                                                    executionVertex.fail(new Throwable()));
+                        },
+                        mainThreadExecutor)
+                .join();
+
+        // wait for task restart and state transition to DEPLOYING
+        while (true) {
```

Comment:
For more accurate and stable testing result, I don't recommend using a while loop to check whether the test completes. In the failover process, we will use the `mainThreadExecutor` and `delayExecutor`, we can do the following things to make all operations be done in one thread(`mainThreadExecutor`).
1.  let `delayExecutor= new ManuallyTriggeredScheduledExecutor()` 
2. trigger `delayExecutor.triggerScheduledTasks()` in `mainThreadExecutor`